src/features/stickyMessage/index.ts

```diff
@@ -0,0 +1,83 @@
+import { Message } from 'discord.js'
+
+import { StickyMessage } from '@prisma/client'
+
+import { Environment } from '../../config.js'
+import { prisma } from '../../prisma.js'
+import { saveCache } from '../../utils/cache.js'
+
+import {
+  ChannelLockType,
+  isChannelLock,
+  lockChannel,
+  unlockChannel,
+} from './lockChannel.js'
+import { resetCooldown, startCooldown } from './messageCooldown.js'
+import { getCounter, resetCounter } from './messageCounter.js'
+
+/**
+ *  Init sticky message memory cache
+ */
+export async function initStickyMessage() {
+  const messages = await prisma.stickyMessage.findMany()
+  for (const message of messages) {
+    saveCache(`sticky-${message.channelId}`, message)
+    resetCounter(message.channelId)
+    startCooldown(message.channelId)
+  }
+}
+
+/**
+ * Pushes the sticky message to the bottom of the channel.
+ * @param {Message} message - The message object representing the triggering message.
+ * @param {StickyMessage} stickyMessage - The sticky message to be pushed to the bottom of the channel.
+ * @returns {Promise<void>} A promise that resolves when the sticky message is successfully pushed to the bottom.
+ */
+export async function pushMessageToBottom(
+  message: Message,
+  stickyMessage: StickyMessage,
+): Promise<void> {
+  // lock channel
+  lockChannel(message.channelId)
+
+  // get old message for delete
+  const oldMessage = await message.channel.messages.fetch(
+    stickyMessage.messageId,
+  )
+
+  await oldMessage.delete()
+
+  const newMessage = await message.channel.send({
+    content: stickyMessage.message,
+  })
+
+  // update sticky message with new id
+  const updatedMessage = await prisma.stickyMessage.update({
+    data: {
+      messageId: newMessage.id,
+    },
+    where: {
+      channelId: message.channelId,
+    },
+  })
```

Comment:
what happens if the code in lines 44–62 fails, will the channel get unlocked automatically somehow, or will the channel will get stuck in locked state?

usually the pattern for safely working with locks in JS/TS is like this:

```ts
lockSomething()
try {
  // ...
} finally {
  unlockSomething()
}
```

i think the lock+cooldown logic can be better represented with a state machine (e.g. [XState](https://xstate.js.org/)) so there is less room for things to go wrong

<img width="642" alt="image" src="https://github.com/creatorsgarten/kaogeek-discord-bot/assets/193136/de9f5527-ce2b-4608-9c8f-77122d0bd524">

…but maybe that can be done later